compiler/luci/pass/src/FoldShapePass.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FoldShapePass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+namespace
+{
+
+template <loco::DataType OutType> luci::CircleConst *folding(luci::CircleShape *shape)
+{
+  auto input_node = loco::must_cast<luci::CircleNode *>(shape->input());
+  auto name = input_node->name();
+  auto shape_status = input_node->shape_status();
+  auto rank = input_node->rank();
+  assert(name.length() > 0);
+  if (shape_status != luci::ShapeStatus::VALID)
+    return nullptr;
+  if (rank <= 0)
+    return nullptr;
+
+  auto folded_shape = input_node->graph()->nodes()->create<luci::CircleConst>();
+  folded_shape->name(name + "_ConstShape");
+  folded_shape->dtype(OutType);
+  folded_shape->rank(1);
+  folded_shape->dim(0).set(rank);
+  luci::add_origin(folded_shape, luci::get_origin(shape));
+
+  folded_shape->size<OutType>(rank);
+  for (uint32_t i = 0; i < rank; i++)
+    folded_shape->at<OutType>(i) = input_node->dim(i).value();
```

Comment:
what happens when dim is unknown? is it acceptable?

Reply:
`value()` is `uint32_t` but may be different with `OutType`

Reply:
> what happens when dim is unknown? is it acceptable?

Thank you for pointing this out!
Since it is not acceptable as you said, I'll update to check if the `dim` is kwown or not 😄

Reply:
> `value()` is `uint32_t` but may be different with `OutType`

Yes, it is bit different now.. but I'm not sure which is the proper way
For now, I just implemented it as follow the default `out_type` value of `tf.shape`, which is signed-32bit integer [[link](https://www.tensorflow.org/api_docs/python/tf/shape)]

Is it better to match as unsigned-32 as follwing the data type of `value()`?

Reply:
You made this PR with `if (fold_shape<loco::DataType::S32>(shape))` call.
Why did you use `S32` ?

Reply:
> Why did you use S32 ?

First of all, my intend was following the default `out_type` of `tf.shape` (`S32`)

But, IIUC, the `out_type` of `CircleShape` can be designated by the user
So I tried to use `shape->out_type()` like below, but it was failed 😥

```cpp
auto out_type = shape->out_type();
if (fold_shape<out_type>(shape))
```

Therefore I made this PR with `S32` first, and added `TODO Support more data types` comment

If there is anything I missed, please let me know :)

Reply:
> TODO Support more data types

what types are you planning to support after this PR?

do you know what output types exist for `Shape` Op?

Reply:
> what types are you planning to support after this PR?

For now, I planned to support `S64` only, and the details are as follow

> do you know what output types exist for Shape Op?

I guess that the `Shape` supports `S32` and `S64`, due to the following statement from [tf.shape](https://www.tensorflow.org/api_docs/python/tf/shape)

> The specified output type of the operation (int32 or int64).

But I'm not sure that these two types cover all the types that supported...
This is the reason why I remained it as a `TODO` item